wakuv2/common/filter.go

```diff
@@ -235,23 +235,24 @@ func (fs *Filters) NotifyWatchers(recvMessage *ReceivedMessage) bool {
 	}
 
 	for _, watcher := range candidates {
-		matched = true
+		// Messages are decrypted successfully only once
 		if decodedMsg == nil {
 			decodedMsg = recvMessage.Open(watcher)
 			if decodedMsg == nil {
 				log.Debug("processing message: failed to open", "message", recvMessage.Hash().Hex(), "filter", watcher.id)
+				continue
```

Comment:
what do you mean by collision here?

it is not expected for 2 different cases to use same `contentTopic+pubsubTopic` , because it would just lead to un-necessary receiving of messages and ultimately getting dropped in application.

Reply:
ah...you mean more than 1 filter per `contentTopic+pubsubTopic`. 

but why would there be more than 1 filter for same set...that doesn't seem right. there can be more than 1 watchers per filter though that want the messages received on it.

Reply:
> but why would there be more than 1 filter for same set...that doesn't seem right

That's because there might be collisions. `contentTopic` is calculated as 4 last bytes of hash of the actual topic. So it is very possible to have different topics sharing same `contentTopic`.

https://github.com/status-im/status-go/blob/063756b4ed401928044baa44f9c7f62b4152b525/wakuv2/common/topic.go#L55-L64

If we would want to prevent it, we would have to keep the whole topic. But then we'll loose privacy.

Reply:
ah right, Thanks for the clarification!

Reply:
> @igor-sirotin: hmm, Isn't (in theory) a contentTopic+pubsubTopic collision is possible? Then we should try to decode the message with a different watcher, because it might have different keys.

Is this question referring to the `// Messages are decrypted successfully only once` comment or to the usage of the `continue` in this piece of code?

If it is the former, encrypted messages can be decrypted with only a single key. I havent reviewed the code throughly but assuming we can have multiple watchers, as long as they have the same key, they can reuse the `decodedMsg` that will contain the decrypted message (Decrypting is considered an expensive operation) . Later `watcher.MatchMessage` will determine if the decrypted message is used by the filter by comparing either the public key of the message or the symmetric key, since comparing bytes is cheap.

```go
	for _, watcher := range candidates {
		// Messages are decrypted successfully only once  <--------------
		if decodedMsg == nil {
			decodedMsg = recvMessage.Open(watcher)
			if decodedMsg == nil {
				log.Debug("processing message: failed to open", "message", recvMessage.Hash().Hex(), "filter", watcher.id)
				continue
			}
		}

		if watcher.MatchMessage(decodedMsg) {   <------------------------------
```

If the question refers to the usage of `continue`, do note that we are running a loop that iterates over candidate watchers. If it couldn't decrypt with the watcher from the current iteration, then it will try on the next iteration, etc.

Reply:
> _encrypted messages can be decrypted with only a single key._

Indeed, didn't think of this. 🤦  
It all makes sense now, thanks!